### ISSUE TITLE
Feedback from Code Review 30/05/2024 

### ISSUE THREAD

**Issue by LennartPurucker** [schaeferbasti/FE_AutoML#1]
Heyho, 

here is a list of things that could improve the code:

- For our experiments, consider setting the range for the number of estimators for LGBM to 200-5000 [here](https://github.com/schaeferbasti/FE_AutoML/blob/main/src/amltk/classifiers/Classifiers.py#L120-L148)
- For classification, let us switch to ROC AUC from [Acc](https://github.com/schaeferbasti/FE_AutoML/blob/main/src/amltk/main.py#L112-L117)
- Double check that your code and models know which features a categorical. I am not sure about the [code](https://github.com/schaeferbasti/FE_AutoML/blob/main/src/amltk/datasets/Datasets.py#L99), but if the data is always a dataframe, then the model might now what is a category and what not. If not, you might be missing this, and the models treat categorical features as numerical. Also, double-check that all the feature generators keep the data frame with the categorical information. 
- Is this the correct Feature Generator for AutoGluon? [this](https://github.com/schaeferbasti/FE_AutoML/blob/main/src/amltk/feature_engineering/AutoGluon.py#L9) I thought we would need to use the AutoMLPipelineFeatureGenerator. On that note, I am also not sure how useful this is as AutoGluon has model-specific preprocessing, e.g., for LightGBM [here](https://github.com/autogluon/autogluon/blob/master/tabular/src/autogluon/tabular/models/lgb/lgb_model.py#L280C9-L280C20), which is for LightGBM basically no feature engineer at all. 
- I noticed that you are [imputing/filling nan values for preprocessing](https://github.com/schaeferbasti/FE_AutoML/blob/main/src/amltk/datasets/Datasets.py#L80). It is quite important for many tree-based models like LighGBM not to do this. Maybe consider saving which cells of the table were nan and setting them as nan before returning the new data. But also, at this time, I think it is okay to ignore this and go with imputing. 
- [Here](https://github.com/schaeferbasti/FE_AutoML/blob/main/src/amltk/feature_engineering/Autofeat.py#L12-L14), you are applying imputing with fit_transform to training and test data separately. You need to adjust this such that test data is only transformed but not used for fitting. Otherwise, you leak information from test data into your eval, likely increasing overfitting or producing a distribution shift from training to test data. 
- For AutoFeat, if we have more transformations than [these](https://github.com/schaeferbasti/FE_AutoML/blob/main/src/amltk/feature_engineering/Autofeat.py#L18) then consider adding them.
- Methods like OpenFE and Autofeat have "steps". We should think about how to set these values to be fair across the different methods. E.g., all the default? all just one step? Ideally, all methods would have a time limit, but most of them failed to implement this.  
- Is the pipeline for [H2O](https://github.com/schaeferbasti/FE_AutoML/blob/main/src/amltk/feature_engineering/H2O.py#L22-L31) from a tutorial or a default value? It looks like this does not generate it automatically, but we would need to generate them. In that case, it is not really an automated feature engineering method, and we can ignore it.  
- [This](https://github.com/schaeferbasti/FE_AutoML/blob/main/src/amltk/feature_engineering/MLJAR.py#L37) sounds a lot like CatBoost's inner method. Do you know what MLJAR does exactly and how? 
- The [sklearn features](https://github.com/schaeferbasti/FE_AutoML/blob/main/src/amltk/feature_engineering/Sklearn.py#L8) are also more manual than automatically generated. Thus, I would say we should not include that in our experiments. 

The code base is very clean and easy to read, well done! 
I think we mostly need to decide what and how we want to benchmark automated feature engineering methods for the sake of the project.

**Comment by schaeferbasti** [comment]
Hi, thank you for your feedback. Following your recommendations, I added the following things:

- Changed the range of n_estimators to the proposed values setting the default to 200
- Switched the Accuracy Metric to `roc_auc`
- The data is always a dataframe
- The IdentityFeatureGenerator is just a dummy, I changed it to the AutoMLPipelineFeatureGenerator, which is supposed to handle tabular data well
- I removed the preprocessing, since it was only necessary for the old classifier
- Even if the preprocessing is not active anymore, I changed the method so that the preprocessor is only fitted on the training data
- I added all possible transformations for autofeat
- I think using a time limit for all methods would be fair, auto feat is usually taking a lot of time. But as you mentioned correctly, OpenFE and autofeat did not implement a timeout. I see two possibilities: 1. Try to measure the time of different configurations and set the values for autofeat, so that they use approximately the same time. 2. Use defaults and live with the inequality. I implement a method measuring the time to keep track of the time and decide of reasonable values for the configuration of the different methods.
- H2O is a bit like autofeat, we need to define the transformations. The disadvantage is, that you need to choose specifically on which column you want to apply which transformation. The advantage is, that is has a lot more possibilities (as a groupby, …)
- This is the process of computing the golden features:
	•	Generate all possible unique pairs of original features.
	•	If there is more than 250,000 pairs then subsample them randomly to 250,000.
	•	For each pair of features construct a new feature with substract or division operators.
	•	Based on the new feature train a Decision Tree with max_depth = 3 (using only one feature).
	•	For training there are used up to 2,500 samples randomly selected from the dataset. The same for testing, also up to 2,500 samples randomly selected.
	•	There is computed a score on test samples for each feature. The score is logloss metric for classification tasks, and mean squared error for regression tasks.
	•	Newly generated features are sorted based on the score (the lower score the better).
	•	As a Golden Features there are selected new features with smallest score values and are inserted into the training data.
	•	The number of Golden Features selected depends on number of original features. It is 5% of original features number, but not less than 5 features and not more than 50 features.
- I removed sklearn from the methods.

**Comment by LennartPurucker** [comment]
Great!

> I think using a time limit for all methods would be fair, auto feat is usually taking a lot of time. But as you mentioned correctly, OpenFE and autofeat did not implement a timeout. I see two possibilities: 1. Try to measure the time of different configurations and set the values for autofeat, so that they use approximately the same time. 2. Use defaults and live with the inequality. I implement a method measuring the time to keep track of the time and decide of reasonable values for the configuration of the different methods.

I am in favor of 2 for now. Fixing this for the other methods is impossible or infeasible for our goal here. 
We should highlight this in our results but show the time it takes to generate the features as well. 
We can then mention this in the discussion. 

> H2O is a bit like autofeat, we need to define the transformations. The disadvantage is, that you need to choose specifically on which column you want to apply which transformation. The advantage is, that is has a lot more possibilities (as a groupby, …)

In that case, I would exclude it as well. To me, H2O also does not do clever pruning, which essentially would leave all the feature engineering to us. But good to know that they support it and have a framework for feature transformations. 

> This is the process of computing the golden features: [...]

I see. This is somewhat similar to OpenFE's feature boosting. It's very interesting, but given the limited transformation space, it's also not too strong. I'm looking forward to seeing how well this performs. Thank you for the recap!
